### ISSUE TITLE
Connection abruptly terminated by the client

### ISSUE THREAD

**Issue by IvanNardi** [private-octopus/quicdoq#25]
Hi
I am doing some trivial tests, and I found that the QUIC connection seems to be abruptly terminated by the client.
The issue is quite evident: the server sends the DNS reply multiple times, receiving an ICMP Port Unreachable error after the second reply: the client missed sending some ACKs or doing proper teardown of the sessions?
How to reproduce:
*) ./quicdoq_app -p 784 -d 8.8.8.8
*) ./quicdoq_app ::
No errors reported in the terminal.
Am I missing something obvious here?

[doq_wrong_termination.zip](https://github.com/private-octopus/quicdoq/files/5612491/doq_wrong_termination.zip)

**Comment by huitema** [comment]
For the client, you are sending packets to the undefined address `::`, i.e., `0::0`. The server is listening at the loopback address, `::1`.

**Comment by IvanNardi** [comment]
Thanks for the reply.
The client option is clearly wrong, my fault... but even if I use "./quicdoq_app ::1", I get the same result
By the way, according to netstat the server, launched via "./quicdoq_app -p 784 -d 8.8.8.8" is listening on ::

```
ivan@latitude-E6540:~/svnrepos/quicdoq(master)$ sudo netstat -natup  |grep quic
udp        0      0 0.0.0.0:784             0.0.0.0:*                           31149/./quicdoq_app 
udp6       0      0 :::784                  :::*                                31149/./quicdoq_app 
```
```

ivan@latitude-E6540:~/svnrepos/quicdoq(master)$ SSLKEYLOGFILE=~/pico.txt ./quicdoq_app ::1 
Query #0 completes after 12554us with code 2
Parsed 56 bytes out of 56:
{ "ID":0, "QR":1, "Opcode":0, "AA":0,
"TC":0, "RD":1, "RA":1, "AD":0, "CD":0, "RCODE":0,
"QDCOUNT":1, "ANCOUNT":1, "NSCOUNT":0, "ARCOUNT":1,
"QNAME": "example.com.", "QTYPE":1, "QCLASS":1,
"answerRRs": [
{ "NAME": "example.com.",
"TYPE":1, "CLASS":1, "TTL":21133,
"RDATAHEX": "5DB8D822"}],
"additionalRRs": [
{ "NAME": ".",
"TYPE":41, "CLASS":512, "TTL":0,
"RDATAHEX": ""}]}
```

Same behaviour: QUIC handshake ok, DNS requet send, DNS reply send the first time, the client prints the reply, DNS reply retrasmitted multiple times getting ICMP errors
New pcap attached (keys embedded): [doq_wrong_termination_b.zip](https://github.com/private-octopus/quicdoq/files/5612886/doq_wrong_termination_b.zip)

![Screenshot from 2020-11-29 22-40-14](https://user-images.githubusercontent.com/12729895/100554297-1a234600-3294-11eb-9531-0e3c0d6d7f30.png)



Thanks for your time

**Comment by huitema** [comment]
Netstat says:
```
udp6       0      0 :::784                  :::*
``` 
which means, listening to "::/0", all addresses. But the IPv6 layer will not deliver a packet sent to ::.

From your logs, the client does receive the response, and closes immediately. The server does not receive a close request from the client, so it repeats packets until giving up on the connection. I think two fixes are needed:

1) On the server, abort the connection immediately if the client address is unreachable
2) On the client, perform a gracious disconnect.

**Comment by huitema** [comment]
@IvanNardi the behavior should be fixed in PR #26
